tpm.go

```diff
@@ -0,0 +1,937 @@
+//
+// If ekCertDataReader is nil, this function will attempt to obtain the endorsement key certificate for the TPM and then download the
+// parent certificates, which requires network access. If this fails, a EkCertVerificationError error will be returned. If network
+// access is unavailable and there is no existing EK certificate blob created previously by FetchAndSaveEkCertificateChain,
+// SaveEkCertificateChain or EncodeEkCertificateChain, then ConnectToDefaultTPM must be used instead.
+//
+// If verification of the endorsement key certificate fails, a EkCertVerificationError error will be returned. If the endorsement
+// key certificate data was supplied by the caller, this might mean that the data is invalid and should be recreated.
+//
+// In order for the TPM to prove it is the device for which the endorsement key certificate was issued, an endorsement key is
+// required. If the TPM doesn't contain a valid persistent endorsement key at the expected location (eg, if ProvisionTPM hasn't been
+// executed yet), this function will attempt to create a transient endorsement key. This requires knowledge of the endorsement
+// hierarchy authorization value, which will be empty on a newly cleared device. If there is no object at the persistent endorsement
+// key index and creation of a transient endorement key fails, ErrProvisioning will be returned.
+//
+// If the TPM cannot prove it is the device for which the endorsement key certificate was issued, a TPMVerificationError error will be
+// returned. This can happen if there is an object at the persistent endorsement key index but it is not the object for which the
+// endorsement key certificate was issued, and creation of a transient endorsement key fails because the correct endorsement hierarchy
+// authorization value hasn't been provided.
+func SecureConnectToDefaultTPM(ekCertDataReader io.Reader, endorsementAuth []byte) (*TPMConnection, error) {
+	tpm, err := connectToDefaultTPM()
+	if err != nil {
+		return nil, err
+	}
+	tpm.EndorsementHandleContext().SetAuthValue(endorsementAuth)
+
+	succeeded := false
+	defer func() {
+		if succeeded {
+			return
+		}
+		tpm.Close()
+	}()
+
+	t := &TPMConnection{TPMContext: tpm}
+
+	var certData *ekCertData
+	if ekCertDataReader != nil {
+		// Unmarshal supplied EK cert data
+		if err := tpm2.UnmarshalFromReader(ekCertDataReader, &certData); err != nil {
+			return nil, EkCertVerificationError{fmt.Sprintf("cannot unmarshal supplied EK certificate data: %v", err)}
+		}
+		if len(certData.Cert) == 0 {
+			// The supplied data only contains parent certificates. Retrieve the EK cert from the TPM.
+			if cert, err := readEkCertFromTPM(tpm); err != nil {
+				return nil, EkCertVerificationError{fmt.Sprintf("cannot obtain endorsement key certificate from TPM: %v", err)}
+			} else {
+				certData.Cert = cert
+			}
+		}
+	} else {
+		data, err := fetchEkCertificateChain(tpm, false)
```

Comment:
what's the advantage of having this path here, vs asking user to call FetchAndSaveEkCertificateChain before, it feels a bit too magical this involving the network implicitly here

Reply:
I can probably remove that - examples/verify-tpm in the original repo does use it, but there probably isn't a reason for it to be there.